ceno_rt/src/mmio.rs

```diff
@@ -0,0 +1,36 @@
+//! Memory-mapped I/O (MMIO) functions.
+
+use rkyv::{Portable, api::high::HighValidator, bytecheck::CheckBytes, rancor::Failure};
+
+use core::slice::from_raw_parts;
+
+extern "C" {
+    /// The address of this variable is the start of the hints ROM.
+    ///
+    /// It is defined in the linker script.
+    static _hints_start: u8;
+    static _hints_end: u8;
+}
+
+static mut NEXT_HINT_LEN_AT: *const u8 = &raw const _hints_start;
+
+pub fn read_slice<'a>() -> &'a [u8] {
+    unsafe {
+        let len: u32 = core::ptr::read(NEXT_HINT_LEN_AT as *const u32);
+        NEXT_HINT_LEN_AT = NEXT_HINT_LEN_AT.add(4);
+
+        let hints_region = {
+            let total_length = (&raw const _hints_end).offset_from(&_hints_start) as usize;
+            from_raw_parts(&_hints_start, total_length)
+        };
```

Comment:
cannot understand here.. 

i guess `read_slice` will be called multi time, so `_hints_start` and `_hints_end` (or one of them) are like cursors, should not be const, they should be changed somewhere.. but i cannot find where they are changed..

Reply:
It's slightly more complicated.

In principle we could construct the slice directly like:
```rust
pub fn read_slice<'a>() -> &'a [u8] {
    unsafe {
        let len: u32 = core::ptr::read(NEXT_HINT_LEN_AT);
        NEXT_HINT_LEN_AT = NEXT_HINT_LEN_AT.add(4);
        &from_raw_parts(&_hints_start, len as usize)
    }
}
```

I hope that variant is easy to understand?

Why do we bother with the more complicated version?

Because `len` is a value that we get given by the prover, so we don't rust it!  So instead we construct the `hints_region`, which is constant and always the same.  Then we return a sub-slice via `&hints_region[..len as usize]`.  That subslicing makes sure to catch any shenanigans a prover might want to play.  (The one I can think of is trying to give a `len` longer than the hints region, but there might be others I'm unaware of.)

As you recognised we compute `hints_region` purely from statically know inputs, and it's always the same.  The only reason it's not declared static is because Rust won't let us.  (But I hope that the optimiser in LLVM is smart enough to recognise that it doesn't need to re-compute.)

Briefly talking about cursors:

You can think of `_hints_start` and `_hints_end` as constant cursors that define the `hints_region`.

`NEXT_HINT_LEN_AT` and `len` are like mutable cursors.  We use `NEXT_HINT_LEN_AT` to find the current value of `len`.

Hope that is clearer?  I'm working making a comment in the code to explain this, too.  (But with some small code changes to make the comment simpler.)